### PR TITLE
CLN: cleanups in DataFrame._reduce

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx
- [ ] tests added / passed
- [ ] passes `black pandas`
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

Broken off from non-CLN work in the vicinity.

### CODE REVIEW

File: pandas/core/frame.py [hunk 1]
```diff
@@ -8684,19 +8676,17 @@ def blk_func(values):
             with np.errstate(all="ignore"):
                 result = func(values)
 
-        if is_object_dtype(result.dtype):
+        if filter_type == "bool" and notna(result).all():
+            result = result.astype(np.bool_)
```
Comment (jreback) [inline pandas/core/frame.py:8680]: can we just call coerce_to_dtypes?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/frame.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/frame.py:8680]: coerce_to_dtypes is about casting back to our `self.dtypes`, not to bool.

Comment (jreback) [reply pandas/core/frame.py:8680]: hmm this is very messy (not your change), but what this is doing. better to overhaul no?

Comment (jbrockmendel) [reply pandas/core/frame.py:8680]: This is specific to any/all, making sure we have bool-dtype.

There are _lots_ of things in _reduce that need an overhaul.  This one doesn't bother me.

**Review by jreback — APPROVED** [review]

**Pull request merged**